src/gui/creds/oauth.cpp

```diff
@@ -0,0 +1,125 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QDesktopServices>
+#include <QNetworkReply>
+#include <QTimer>
+#include "account.h"
+#include <json.h>
+#include "creds/oauth.h"
+#include "theme.h"
+
+
+namespace OCC
+{
+
+OAuth::~OAuth()
+{
+}
+
+static void httpReplyAndClose(QTcpSocket *socket, const char *code, const char *html)
+{
+    socket->write("HTTP/1.1 ");
+    socket->write(code);
+    socket->write("\r\nContent-Type: text/html\r\nConnection: close\r\nContent-Length: ");
+    socket->write(QByteArray::number(qstrlen(html)));
+    socket->write("\r\n\r\n");
+    socket->write(html);
+    socket->disconnectFromHost();
+}
+
+void OAuth::start()
+{
+    // Listen on the socket to get a port which will be used in the redirect_uri
+    if (!_server.listen(QHostAddress::LocalHost)) {
+        emit result(NotSupported, QString());
+        return;
+    }
+
+    if (!openBrowser())
+        return;
+
+    emit result(Waiting);
+
+    QObject::connect(&_server, &QTcpServer::newConnection, this, [this] {
+        while (QTcpSocket *socket = _server.nextPendingConnection()) {
+            QObject::connect(socket, &QTcpSocket::disconnected, socket, &QTcpSocket::deleteLater);
+            QObject::connect(socket, &QIODevice::readyRead, this, [this, socket]{
+                QByteArray peek = socket->peek(qMax(socket->bytesAvailable(), 4000LL)); //The code should always be within the first 4K
+                if (peek.indexOf('\n') < 0)
+                    return;  // wait until we find a \n
+                QRegExp rx("^GET /\\?code=([a-zA-Z0-9]+)[& ]"); // Match a  /?code=...  URL
+                if (rx.indexIn(peek) != 0) {
+                    httpReplyAndClose(socket, "404 Not Found", "<html><head><title>404 Not Found</title></head><body><center><h1>404 Not Found</h1></center></body></html>");
+                    return;
+                }
+
+                // TODO: add redirect to the page on the server
+                httpReplyAndClose(socket, "200 OK", "<h1>Login Successfull</h1><p>You can close this window.</p>");
+
+                QString code = rx.cap(1); // The 'code' is the first capture of the regexp
+
+                QUrl requestToken(_account->url().toString()
+                    + QLatin1String("/index.php/apps/oauth2/api/v1/token?grant_type=authorization_code&code=")
+                    + code
+                    + QLatin1String("&redirect_uri=http://localhost:") + QString::number(_server.serverPort()));
```

Comment:
This, and the following code, is pretty similar to the "use refresh token to get an access token" code further down. Can they be merged, and if not, at least be moved to the same file?

Also, what does this second redirect_url even do? It never gets used, right?